build_runner/lib/src/generate/build_impl.dart

```diff
@@ -411,12 +411,38 @@ class BuildImpl {
       SingleStepReader reader, AssetWriterSpy writer) async {
     // Reset the state for each output, setting `wasOutput` to false for now
     // (will set to true in the next loop for written assets).
+    //
+    // Also updates the `inputs` set for each output, and the `outputs` sets for
+    // all inputs.
     for (var output in declaredOutputs) {
-      (_assetGraph.get(output) as GeneratedAssetNode)
+      var node = _assetGraph.get(output) as GeneratedAssetNode;
+      node
         ..needsUpdate = false
         ..wasOutput = false
         ..lastKnownDigest = null
         ..globs = reader.globsRan.toSet();
+
+      // Update dependencies that don't exist any more.
+      var removedInputs = node.inputs.difference(reader.assetsRead);
+      node.inputs.removeAll(removedInputs);
+      for (var input in removedInputs) {
+        // TODO: special type of dependency here? This means the primary input
```

Comment:
IMO it's fine to always treat the primary input like it was read. The edge cases shouldn't break anything, at worst builds are slightly less optimal

Reply:
ya, thats pretty much my same line of thinking for now.